flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java

```diff
@@ -522,6 +527,7 @@ protected ComponentMainThreadExecutor getMainThreadExecutor() {
     protected void failJob(Throwable cause) {
         incrementVersionsOfAllVertices();
         executionGraph.failJob(cause);
+        getTerminationFuture().thenRun(() -> archiveGlobalFailure(cause));
```

Comment:
I think we should check here that we are indeed in the `FAILED` state. What can also happen is that the user cancels the job in between.

Reply:
Here, I don't understand fully: You're referring to the case where a failure happens, the user cancels the job while the failure handling is done and the `failJob` method might be called while being in a `CANCELED`/`CANCELLING` state?For this case, I would still think that we should archive the exception because the users intervention happened after the exception happened.

Reply:
I mean the case that a failure happens, the job goes into the `FAILING` state and tries to cancel the tasks and now the user cancels the job because it takes too long for him. Then the job will go into the `CANCELING` state which will result to the `CANCELED` state once all tasks have terminated.

I think you are right that we should still record the failure cause.